compiler-core/src/error.rs

```diff
@@ -3731,21 +3729,1020 @@ and explain your usecase for this pattern, and how you would expect it to behave
+                                text: "A generic type must have at least a generic parameter.".into(),
+                                hint: Some("If a type is not generic you should omit the `()`.".into()),
+                                level: Level::Error,
+                                location: Some(Location {
+                                    label: Label {
+                                        text: Some("I was expecting generic parameters here".into()),
+                                        span: location,
+                                    },
+                                    path: path.clone(),
+                                    src: src.clone(),
+                                    extra_labels: vec![],
+                                }),
+                            },
+
+                        ParseErrorType::UnknownAttributeRecordVariant =>
+                            Diagnostic {
+                                title,
+                                text: "".into(),
+                                hint: Some("Did you mean `@deprecated`?".into()),
+                                level: Level::Error,
+                                location: Some(Location {
+                                    label: Label {
+                                        text: Some("This attribute cannot be used on a variant.".into()),
+                                        span: location,
+                                    },
+                                    path: path.clone(),
+                                    src: src.clone(),
+                                    extra_labels: vec![],
+                                }),
+                            },
+
+                        ParseErrorType::IncorrectImportModuleSeparator { module, item } =>
+                            Diagnostic {
+                                title,
+                                text: [
+                                    "Perhaps you meant one of:".into(),
+                                    "".into(),
+                                    format!("    import {module}/{item}"),
+                                    format!("    import {module}.{{item}}"),
+                                ].join("\n"),
+                                hint: None,
+                                level: Level::Error,
+                                location: Some(Location {
+                                    label: Label {
+                                        text: Some("I was expecting either `/` or `.{` here.".into()),
+                                        span: location,
+                                    },
+                                    path: path.clone(),
+                                    src: src.clone(),
+                                    extra_labels: vec![],
+                                }),
+                            },
+                    };
+
+                vec![diagnostic]
```

Comment:
There's a lot of duplication here, making it verbose and hard to see the bits that change. Please extract just the variable bits please 🙏

Reply:
It's the same as what we do for the other type errors, should I do the same for those? I don't think we've never had a problem for those. I went with this approach and removed the code abstracting the changing bits because for this new error I needed to add some extra labels so everything would have been turned into a triple with a final empty list except for this new error:
```diff
- (title, lines)
+ (title, lines, [])
```
For some reason it felt worse than just building the diagnostic directly, but I can revert it and go for triples instead!

Reply:
Also to play nicely with the abstracting bit of code we built hints by hand as a string mixed with the other error message instead of using the proper field of the `Diagnostic` struct:
```rs
let (text_lines, label_text) = (
  vec!["Hint: wibble wobble", "Error!", "See: ..."],
  "label"
);

Diagnostic {
  text: text_lines.join("\n"),
  hint: None,
  ..
}
```
Instead of what would make the best use of the diagnostic fields:
```rs
let (text_lines, hint, label_text) = (
  vec!["Error!", "See: ..."],
  Some("wibble wobble"),
  "label",
);

Diagnostic {
  text: text_lines.join("\n"),
  hint,
  ..
}
```

Reply:
For me the goal is to be able to clearly see what each match clause does, so having a tuple or that struct would be greatly helpful! Right now I can't identify what the important bits are from reading unless I take note because the majority of the code is the same in each clause.